### PR TITLE
Run e2e2 Tests

### PR DESCRIPTION
**Pull request by nils-wisiol**
TODO:

- [x] change branch name of desec-stack to `main` after merging desec-io/desec-stack#641
- [x] do not remove cache
- [x] fix easyrsa usage (https://github.com/OpenVPN/easy-rsa/issues/717)

### CODE REVIEW

**Comment by nils-wisiol** [comment]
Ready for review with the exception of the cache removal. Ideas on how to deal with the cache are welcome - I confirmed that with the cache enabled, the tests won't pass. (But only the rrset tests - rr tests work fine ...)

File: docker-compose.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ services:
       - DESEC_NS_NAME
     volumes:
       - certs:/etc/certs
-      - ./certs/entrypoint.sh:/usr/local/bin/entrypoint.sh
+      - ./dox-certs/entrypoint.sh:/usr/local/bin/entrypoint.sh
```
Comment (peterthomassen) [inline docker-compose.yml:10]: I think we should keep the service name and the directory name in sync.

File: docker-compose.yml [hunk 2]
```diff
@@ -27,10 +27,10 @@ services:
     - DESEC_NS_SIGNALING_DOMAIN_SOA_RNAME
     - DESEC_NS_COOKIES_SECRET
     networks:
-      middle:
+      fmiddle:
         ipv4_address: 10.16.2.2
-      rear:
-      rearreplication:
+      frear:
+      frearreplication:
```
Comment (peterthomassen) [inline docker-compose.yml]: While I don't feel very strongly, I think that "telling names" would be better for readability/maintainability (such as `ns_replication`).

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        ln -s ../desec-ns/dox-certs
+        ln -s ../desec-ns/ns
+        ln -s ../desec-ns/openvpn-client
+        ln -s ../desec-ns/replicator
+      working-directory: ./desec-stack
+
+    - name: Create necessary symlinks in desec-ns
+      run: |
+        ln -s ../desec-stack/openvpn-server
+      working-directory: ./desec-ns
```
Comment (peterthomassen) [inline .github/workflows/test.yml]: Add comment why this is necessary

File: docker-compose.connect-stack.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      - DESECSTACK_VPN_SERVER=openvpn-server
+    networks:
+      vpnconnect:
+  openvpn-server:
+    networks:
+      vpnconnect:
+  replicator:
+    environment:
+      - DESECSTACK_VPN_SERVER=desec.$DESECSTACK_DOMAIN
+      - DESEC_NS_E2E2=test mode
```
Comment (peterthomassen) [inline docker-compose.connect-stack.yml]: Does the string value mean anything? (Else, `1` would be a more canonical value.)

File: docker-compose.connect-stack.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  test-e2e2:
+    environment:
+      - DESECSTACK_E2E2_SECONDARY_NS=10.16.0.2
```
Comment (peterthomassen) [inline docker-compose.connect-stack.yml:18]: What does this variable do?

File: replicator/run.py [hunk 6]
```diff
@@ -7,6 +7,7 @@
 import requests
 
 
+testing = os.environ.get('DESEC_NS_E2E2', None) is not None
```
Comment (peterthomassen) [inline replicator/run.py]: Easier: `verify = os.environ.get('DESEC_NS_E2E2') is None`, then use `verify` below. (`get()` defaults to `None`.)

File: vpn-setup.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+TARGET="$(pwd)"
+cd "$(mktemp -d)" || exit
+git clone https://github.com/OpenVPN/easy-rsa.git
```
Comment (peterthomassen) [inline vpn-setup.sh]: consider checking out a specific version/tag/commit

**Review by peterthomassen — COMMENTED** [review]
cool stuff!

File: docker-compose.connect-stack.yml (hunk 5 above)
Comment (nils-wisiol) [reply docker-compose.connect-stack.yml:18]: Defines the target host for e2e2 tests. In desec-stack this is set to nsmaster, here it points to dnsdist of desec-ns.

Comment (nils-wisiol) [reply docker-compose.connect-stack.yml:18]: (Tests also always run against nslord, but with different timing constraints.)

File: docker-compose.yml (hunk 1 above)
Comment (nils-wisiol) [reply docker-compose.yml:10]: done

File: docker-compose.yml (hunk 2 above)
Comment (nils-wisiol) [reply docker-compose.yml]: done, I avoided the `_` though.

File: .github/workflows/test.yml (hunk 3 above)
Comment (nils-wisiol) [reply .github/workflows/test.yml]: done

File: docker-compose.connect-stack.yml (hunk 4 above)
Comment (nils-wisiol) [reply docker-compose.connect-stack.yml]: done

File: replicator/run.py (hunk 6 above)
Comment (nils-wisiol) [reply replicator/run.py]: done

File: vpn-setup.sh (hunk 7 above)
Comment (nils-wisiol) [reply vpn-setup.sh]: done

**Comment by nils-wisiol** [comment]
I've addressed all review comments. Changes to the "small" commits are already squashed, changes to "feat(): run desec-stack's e2e2 tests as CI" are listed as fixups.

**Review by peterthomassen — COMMENTED** [review]
lgtm (modulo cache)

**Comment by nils-wisiol** [comment]
Ready to merge as soon as the tests pass.

**Review by peterthomassen — APPROVED** [review]
lvgtm!

**Pull request merged**